File: clients/windows/src/ChatOS.Core/State/ClientOwnedWorkspaceLoader.cs
using ChatOS.Core.Abstractions;
using ChatOS.Core.Domain;

namespace ChatOS.Core.State;

public sealed record WorkspaceRelationsSnapshot(
    IReadOnlyList<WorkspaceContact> Contacts, IReadOnlyList<WorkspaceConversation> Conversations);

public sealed record ClientOwnedWorkspaceLoadResult(WorkspaceSnapshot Snapshot, string? RemoteError);

// Construct per authenticated account. Publish LoadLocalAsync before awaiting RefreshAsync.
// Never imports or merges remote project entities and never hides local database failures.
// Callers must discard stale authentication/refresh generations before publishing either result.
public sealed class ClientOwnedWorkspaceLoader
{
    private readonly IProjectRegistry _registry;
    private readonly IWorkspaceRelationsService _remote;
    private readonly string _ownerUserId;

    public ClientOwnedWorkspaceLoader(IProjectRegistry registry, IWorkspaceRelationsService remote, string ownerUserId)
    {
        ProjectRegistryValidation.Identifier(ownerUserId, nameof(ownerUserId));
        _registry = registry;
        _remote = remote;
        _ownerUserId = ownerUserId;
    }

    public Task<WorkspaceSnapshot> LoadLocalAsync(string? deviceId, CancellationToken cancellationToken = default) =>
        ComposeAsync(deviceId, new([], []), cancellationToken);

    public async Task<ClientOwnedWorkspaceLoadResult> RefreshAsync(
        string? deviceId, CancellationToken cancellationToken = default)
    {
        WorkspaceRelationsSnapshot relations;
        string? remoteError = null;
        try { relations = await _remote.FetchWorkspaceRelationsAsync(cancellationToken).ConfigureAwait(false); }
        catch (Exception error) when (error is not OperationCanceledException && !cancellationToken.IsCancellationRequested)
        {
            relations = new([], []);
            remoteError = error.Message;
        }
        // Read after the network await so stale remote responses cannot resurrect deleted projects.
        return new(await ComposeAsync(deviceId, relations, cancellationToken).ConfigureAwait(false), remoteError);
    }

    private async Task<WorkspaceSnapshot> ComposeAsync(
        string? deviceId, WorkspaceRelationsSnapshot relations, CancellationToken cancellationToken)
    {
        cancellationToken.ThrowIfCancellationRequested();
        var records = await _registry.ListAsync(_ownerUserId, cancellationToken: cancellationToken).ConfigureAwait(false);
        var conversations = relations.Conversations.Where(c => !c.IsArchived)
            .OrderByDescending(c => c.UpdatedAt).ThenBy(c => c.Id, StringComparer.Ordinal).ToArray();
        var projects = records.Select(record => new WorkspaceProject(
            record.Id, record.Draft.Name, LocalRootUri(record, deviceId), null,
            conversations.FirstOrDefault(c => c.ProjectId == record.Id)?.Id,
            deviceId is null ? null : ProjectContextSnapshot.FromRecord(record, deviceId))).ToArray();
        return new(projects, relations.Contacts, relations.Conversations);
    }

    public static string? LocalRootUri(LocalProjectRecord record, string? deviceId)
    {
        record.Validate();
        if (deviceId is null) return null;
        foreach (var value in new[] { deviceId, record.Draft.WorkspaceId })
        {
            ProjectRegistryValidation.RouteIdentifier(value, "executionTarget");
        }
        var root = $"local://connector/{Uri.EscapeDataString(deviceId)}/{Uri.EscapeDataString(record.Draft.WorkspaceId)}";
        return record.Draft.RelativeRoot.Length == 0 ? root :
            root + "/" + string.Join("/", record.Draft.RelativeRoot.Split('/').Select(Uri.EscapeDataString));
    }
}

File: clients/macos/Sources/ChatOSCore/ClientOwnedWorkspaceLoader.swift
import Foundation

public struct WorkspaceRelationsSnapshot: Sendable, Equatable {
    public let contacts: [WorkspaceContact]
    public let conversations: [WorkspaceConversation]

    public init(contacts: [WorkspaceContact], conversations: [WorkspaceConversation]) {
        self.contacts = contacts
        self.conversations = conversations
    }
}

public protocol WorkspaceRelationsRemoteServicing: Sendable {
    /// Must not fetch or synthesize project entities.
    func fetchWorkspaceRelations() async throws -> WorkspaceRelationsSnapshot
}

public struct ClientOwnedWorkspaceLoadResult: Sendable {
    public let snapshot: WorkspaceSnapshot
    public let remoteError: String?
}

/// Account-bound composition for migrated clients. No implicit import, remote project merge,
/// or fallback to server project authority. UI should publish loadLocal() before awaiting refresh().
/// Callers must discard stale authentication/refresh generations before publishing either result.
public struct ClientOwnedWorkspaceLoader: Sendable {
    private let registry: any ProjectRegistry
    private let remote: any WorkspaceRelationsRemoteServicing
    private let ownerUserID: String

    public init(registry: any ProjectRegistry, remote: any WorkspaceRelationsRemoteServicing, ownerUserID: String) throws {
        try ProjectRegistryValidation.identifier(ownerUserID, field: "ownerUserID")
        self.registry = registry
        self.remote = remote
        self.ownerUserID = ownerUserID
    }

    public func loadLocal(deviceID: String?) async throws -> WorkspaceSnapshot {
        try await compose(relations: .init(contacts: [], conversations: []))
    }

    public func refresh(deviceID: String?) async throws -> ClientOwnedWorkspaceLoadResult {
        let relations: WorkspaceRelationsSnapshot
        let remoteError: String?
        do {
            relations = try await remote.fetchWorkspaceRelations()
            remoteError = nil
        } catch {
            try Task.checkCancellation()
            if error is CancellationError { throw error }
            relations = .init(contacts: [], conversations: [])
            remoteError = error.localizedDescription
        }
        // Read after the network await: a pending remote response cannot resurrect local deletions.
        let snapshot = try await compose(relations: relations)
        return .init(snapshot: snapshot, remoteError: remoteError)
    }

    private func compose(relations: WorkspaceRelationsSnapshot) async throws -> WorkspaceSnapshot {
        try Task.checkCancellation()
        let records = try await registry.list(ownerUserID: ownerUserID, includeInactive: false)
        let conversations = relations.conversations.filter { !$0.isArchived }.sorted {
            $0.updatedAt == $1.updatedAt ? $0.id < $1.id : $0.updatedAt > $1.updatedAt
        }
        let projects = try records.map { record in
            return WorkspaceProject(
                id: record.id, name: record.draft.name, rootPath: record.draft.rootPath,
                latestConversationID: conversations.first { $0.projectID == record.id }?.id,
                projectContext: try ProjectContextSnapshot(record: record)
            )
        }
        return .init(projects: projects, contacts: relations.contacts, conversations: relations.conversations)
    }
}
